File: app/src/main/java/com/omar/retromp3recorder/app/repo/RequestPermissionsRepo.java
package com.omar.retromp3recorder.app.repo;

import com.omar.retromp3recorder.app.utils.OneShot;

import java.util.Set;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Observable;
import io.reactivex.subjects.BehaviorSubject;

@Singleton
public class RequestPermissionsRepo {

    private final BehaviorSubject<OneShot<ShouldRequestPermissions>> state = BehaviorSubject.create();

    @Inject
    public RequestPermissionsRepo(){}

    public Observable<OneShot<ShouldRequestPermissions>> observe() {
        return state;
    }

    public void newValue(ShouldRequestPermissions newValue) {
        state.onNext(new OneShot<>(newValue));
    }

    //region events
    public interface ShouldRequestPermissions { }

    public static class Granted implements ShouldRequestPermissions { }

    public static class Denied implements ShouldRequestPermissions {
        public final Set<String> permissions;

        public Denied(Set<String> permissions) {
            this.permissions = permissions;
        }
    }
    //endregion
}

File: app/src/main/java/com/omar/retromp3recorder/app/common/repo/RequestPermissionsRepo.kt
package com.omar.retromp3recorder.app.common.repo

import com.github.alkurop.ghostinshell.Shell
import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class RequestPermissionsRepo @Inject constructor() {
    private val state: BehaviorSubject<ShouldRequestPermissions> = BehaviorSubject.create()
    fun observe(): Observable<ShouldRequestPermissions> {
        return state
    }

    fun newValue(newValue: ShouldRequestPermissions) {
        state.onNext(newValue)
    }

    sealed class ShouldRequestPermissions {
        object Granted : ShouldRequestPermissions()
        data class Denied(val permissions: Shell<Set<String>>) : ShouldRequestPermissions()
    }
}